### ISSUE TITLE
Add `--with` to CLI to selectively enable extension support "outside" extension commands

### ISSUE THREAD

**Issue by mih** [datalad/datalad#6589]
There are various shortcomings re how extensions can influence the core package: https://github.com/datalad/datalad/issues/5996

However, for DataLad use in a Python session most of them are more of a nuisance than a strict limitation, because users can simply import an extension directly in such a session, enabling the extension to arbitrarily customize the core package.

There is no equivalent in the CLI. In contrast, the CLI tries hard to avoid extension imports that are not needed for execution a specific command -- only once it is clear that an extensions needs to be imported, this is finally done. Consequently, running core commands like `save` or `run` through the CLI will not import any extensions. This, more than anything else, limits the effective impact of extension code.

I propose to implement a `--with` option for the main DataLad CLI entrypoint. If given (could be multiple times), an extension entrypoint for the given name will be loaded, which causes the extension to be imported. A matching configuration setting would allow for configuring a set of extensions to be always loaded unconditionally for a user and/or dataset. This puts extension code into the position to customize behavior and capabilities of core commands, arbitrarily. The name is chosen, because the mechanism resembles a `with patch(....):` context manager use for a DataLad session.

Such a feature would enable use cases like:

- make clone run something other than `git-clone` to establish a repo https://github.com/datalad/datalad-ebrains/issues/21, e.g. 

  `datalad --with ebrains clone https://search.kg.ebrains.eu/instances/e472a8c7-d9f9-4e75-9d0b-b137cecbc6a2`

- make `clone` and `push` deploy necessary credentials to enable or push to special remotes of particular types https://github.com/mih/datalad-mihextras/issues/30

If this acknowledged as a recognized approach to customize datalad, existing internal helpers could be promoted to a more "public" status with a more careful/stable API choice for reliably supporting this kind of "remote decoration".

If this feature would have existed, it would likely have avoided the need to (prematurely) merge the RIA code base into the core package -- which was only needed, because `clone` could not be customized with the necessary pre and post annex-init hooks.

**Comment by bpoldrack** [comment]
Sounds like a sensible compromise re import times to me.

I guess, it would still allow to make the python "experience" very similar to the CLI, by providing a respective helper in the `datalad` (or `datalad.api`) module, like so:

```
import datalad
datalad.with(extensions=[...])
```

**Comment by mih** [comment]
If I were to implement it `datalad.with(extensions=[...])` would be unnecessary, because it would be equivalent to the well-understood `import <whatever_extension_package>`

**Comment by mih** [comment]
I figured that it is not needed to have a dedicated option for this, because we have the config mechanism anyways (however, see #6592). With this a solution is trivial and would be fast after #6591 is merged. For now, I will just leave the respective patch from my dev branch here

```diff
diff --git a/datalad/cli/main.py b/datalad/cli/main.py
index ae98035dc..0b82a6775 100644
--- a/datalad/cli/main.py
+++ b/datalad/cli/main.py
@@ -112,6 +112,11 @@ def main(args=sys.argv):
     if 'datalad.runtime.librarymode' in datalad.cfg:
         datalad.enable_librarymode()
 
+    # load extensions requested by configuration
+    # analog to what coreapi is doing for a Python session
+    from datalad.support.entryproints import load_extensions
+    load_extensions()
+
     if cmdlineargs.change_path is not None:
         from datalad.utils import chpwd
         for path in cmdlineargs.change_path:
diff --git a/datalad/coreapi.py b/datalad/coreapi.py
index ae03ff6c3..0134ca7c4 100644
--- a/datalad/coreapi.py
+++ b/datalad/coreapi.py
@@ -18,6 +18,14 @@ def _generate_func_api():
     """
     from importlib import import_module
 
+    # load extensions requested by configuration
+    import datalad
+    if datalad.get_apimode() == 'python':
+        # only do this in Python API mode, because the CLI main
+        # will have done this already
+        from datalad.support.entryproints import load_extensions
+        load_extensions()
+
     from .interface.base import get_interface_groups
     from .interface.base import get_api_name
 
diff --git a/datalad/interface/common_cfg.py b/datalad/interface/common_cfg.py
index dda6a17bb..0050a63f4 100644
--- a/datalad/interface/common_cfg.py
+++ b/datalad/interface/common_cfg.py
@@ -117,6 +117,20 @@ definitions = {
         'type': bool,
         'default': False,
     },
+    'datalad.extensions.load': {
+        'ui': ('question', {
+               'title': 'DataLad extension packages to load',
+               'text': 'Indicate which extension packages should be loaded '
+                       'unconditionally on CLI startup or on importing '
+                       "'datalad.[core]api'. This enables the "
+                       'respective extensions to customize DataLad with '
+                       'functionality and configurability outside the '
+                       'scope of extension commands. For merely running '
+                       'extension commands it is not necessary to load them '
+                       'specifically'}),
+        'destination': 'global',
+        'default': None,
+    },
     'datalad.externals.nda.dbserver': {
         'ui': ('question', {
                'title': 'NDA database server',
diff --git a/datalad/support/entrypoints.py b/datalad/support/entrypoints.py
index a89782b86..047481d34 100644
--- a/datalad/support/entrypoints.py
+++ b/datalad/support/entrypoints.py
@@ -63,3 +63,30 @@ def iter_entrypoints(group, load=False):
                 ep.name, group, ce)
             continue
     lgr.debug("Done processing entrypoints")
+
+
+def load_extensions():
+    """Load entrypoint for any configured extension package
+
+    Log a warning in case a requested extension is not available, or if
+    a requested extension fails on load.
+
+    Extensions to load are taken from the 'datalad.extensions.load'
+    configuration item.
+    """
+    load_extensions = cfg.get('datalad.extensions.load', get_all=True)
+    if load_extensions:
+        from datalad.utils import ensure_list
+        exts = {
+            ename: eload
+            for ename, _, eload in iter_entrypoints('datalad.extensions')
+        }
+        for el in ensure_list(load_extensions):
+            if el not in exts:
+                lgr.warning('Requested extension %r is not available', el)
+                continue
+            try:
+                exts[el]()
+            except Exception as e:
+                ce = CapturedException(e)
+                lgr.warning('Could not load extension %r: %s', ce)
```
